{"info": {"author": "Bob Green, Alexandr Skurikhin", "author_email": "bgreen@litl.com, a.skurihin@gmail.com", "bugtrack_url": null, "classifiers": ["Development Status :: 5 - Production/Stable", "Intended Audience :: Developers", "License :: OSI Approved :: MIT License", "Natural Language :: English", "Operating System :: OS Independent", "Programming Language :: Python", "Programming Language :: Python :: 3.5", "Programming Language :: Python :: Implementation", "Programming Language :: Python :: Implementation :: CPython", "Topic :: Internet :: WWW/HTTP", "Topic :: Software Development :: Libraries :: Python Modules", "Topic :: Utilities"], "description": "backoff\n=======\n\n.. image:: https://travis-ci.org/litl/backoff.svg?branch=master\n    :target: https://travis-ci.org/litl/backoff?branch=master\n.. image:: https://coveralls.io/repos/litl/backoff/badge.svg?branch=master\n    :target: https://coveralls.io/r/litl/backoff?branch=master\n\n**Function decoration for backoff and retry**\n\nThis module provides function decorators which can be used to wrap a\nfunction such that it will be retried until some condition is met. It\nis meant to be of use when accessing unreliable resources with the\npotential for intermittent failures i.e. network resources and external\nAPIs. Somewhat more generally, it may also be of use for dynamically\npolling resources for externally generated content.\n\nExamples\n========\n\nSince Kenneth Reitz's `requests <http://python-requests.org>`_ module\nhas become a defacto standard for HTTP clients in python, networking\nexamples below are written using it, but it is in no way required by\nthe backoff module.\n\n@backoff.on_exception\n---------------------\n\nThe ``on_exception`` decorator is used to retry when a specified exception\nis raised. Here's an example using exponential backoff when any\n``requests`` exception is raised::\n\n    @backoff.on_exception(backoff.expo,\n                          requests.exceptions.RequestException,\n                          max_tries=8)\n    def get_url(url):\n        return requests.get(url)\n\nThe decorator will also accept a tuple of exceptions for cases where\nyou want the same backoff behavior for more than one exception type::\n\n    @backoff.on_exception(backoff.expo,\n                          (requests.exceptions.Timeout,\n                           requests.exceptions.ConnectionError),\n                          max_tries=8)\n    def get_url(url):\n        return requests.get(url)\n\nIn some cases the raised exception instance itself may need to be\ninspected in order to determine if it is a retryable condition. The\n``giveup`` keyword arg can be used to specify a function which accepts\nthe exception and returns a truthy value if the exception should not\nbe retried::\n\n    def fatal_code(e):\n        return 400 <= e.response.status_code < 500\n\n    @backoff.on_exception(backoff.expo,\n                          requests.exceptions.RequestException,\n                          max_tries=8,\n                          giveup=fatal_code)\n    def get_url(url):\n        return requests.get(url)\n\n\n@backoff.on_predicate\n---------------------\n\nThe ``on_predicate`` decorator is used to retry when a particular\ncondition is true of the return value of the target function.  This may\nbe useful when polling a resource for externally generated content.\n\nHere's an example which uses a fibonacci sequence backoff when the\nreturn value of the target function is the empty list::\n\n    @backoff.on_predicate(backoff.fibo, lambda x: x == [], max_value=13)\n    def poll_for_messages(queue):\n        return queue.get()\n\nExtra keyword arguments are passed when initializing the\nwait generator, so the ``max_value`` param above is passed as a keyword\narg when initializing the fibo generator.\n\nWhen not specified, the predicate param defaults to the falsey test,\nso the above can more concisely be written::\n\n    @backoff.on_predicate(backoff.fibo, max_value=13)\n    def poll_for_message(queue)\n        return queue.get()\n\nMore simply, a function which continues polling every second until it\ngets a non-falsey result could be defined like like this::\n\n    @backoff.on_predicate(backoff.constant, interval=1)\n    def poll_for_message(queue)\n        return queue.get()\n\nJitter\n------\n\nA jitter algorithm can be supplied with the ``jitter`` keyword arg to\neither of the backoff decorators. This argument should be a function\naccepting the original unadulterated backoff value and returning it's\njittered counterpart.\n\nAs of version 1.2, the default jitter function ``backoff.full_jitter``\nimplements the 'Full Jitter' algorithm as defined in the AWS\nArchitecture Blog's `Exponential Backoff And Jitter\n<https://www.awsarchitectureblog.com/2015/03/backoff.html>`_ post.\n\nPrevious versions of backoff defaulted to adding some random number of\nmilliseconds (up to 1s) to the raw sleep value. If desired, this\nbehavior is now available as ``backoff.random_jitter``.\n\nUsing multiple decorators\n-------------------------\n\nThe backoff decorators may also be combined to specify different\nbackoff behavior for different cases::\n\n    @backoff.on_predicate(backoff.fibo, max_value=13)\n    @backoff.on_exception(backoff.expo,\n                          requests.exceptions.HTTPError,\n                          max_tries=4)\n    @backoff.on_exception(backoff.expo,\n                          requests.exceptions.TimeoutError,\n                          max_tries=8)\n    def poll_for_message(queue):\n        return queue.get()\n\nRuntime Configuration\n---------------------\n\nThe decorator functions ``on_exception`` and ``on_predicate`` are\ngenerally evaluated at import time. This is fine when the keyword args\nare passed as constant values, but suppose we want to consult a\ndictionary with configuration options that only become available at\nruntime. The relevant values are not available at import time. Instead,\ndecorator functions can be passed callables which are evaluated at\nruntime to obtain the value::\n\n    def lookup_max_tries():\n        # pretend we have a global reference to 'app' here\n        # and that it has a dictionary-like 'config' property\n        return app.config[\"BACKOFF_MAX_TRIES\"]\n\n    @backoff.on_exception(backoff.expo,\n                          ValueError,\n                          max_tries=lookup_max_tries)\n\nMore cleverly, you might define a function which returns a lookup\nfunction for a specified variable::\n\n    def config(app, name):\n        return functools.partial(app.config.get, name)\n\n    @backoff.on_exception(backoff.expo,\n                          ValueError,\n                          max_value=config(app, \"BACKOFF_MAX_VALUE\")\n                          max_tries=config(app, \"BACKOFF_MAX_TRIES\"))\n\nEvent handlers\n--------------\n\nBoth backoff decorators optionally accept event handler functions\nusing the keyword arguments ``on_success``, ``on_backoff``, and ``on_giveup``.\nThis may be useful in reporting statistics or performing other custom\nlogging.\n\nHandlers must be callables with a unary signature accepting a dict\nargument. This dict contains the details of the invocation. Valid keys\ninclude:\n\n* *target*: reference to the function or method being invoked\n* *args*: positional arguments to func\n* *kwargs*: keyword arguments to func\n* *tries*: number of invocation tries so far\n* *wait*: seconds to wait (``on_backoff`` handler only)\n* *value*: value triggering backoff (``on_predicate`` decorator only)\n\nA handler which prints the details of the backoff event could be\nimplemented like so::\n\n    def backoff_hdlr(details):\n        print (\"Backing off {wait:0.1f} seconds afters {tries} tries \"\n               \"calling function {func} with args {args} and kwargs \"\n               \"{kwargs}\".format(**details))\n\n    @backoff.on_exception(backoff.expo,\n                          requests.exceptions.RequestException,\n                          on_backoff=backoff_hdlr)\n    def get_url(url):\n        return requests.get(url)\n\n**Multiple handlers per event type**\n\nIn all cases, iterables of handler functions are also accepted, which\nare called in turn. For example, you might provide a simple list of\nhandler functions as the value of the ``on_backoff`` keyword arg::\n\n    @backoff.on_exception(backoff.expo,\n                          requests.exceptions.RequestException,\n                          on_backoff=[backoff_hdlr1, backoff_hdlr2])\n    def get_url(url):\n        return requests.get(url)\n\n**Getting exception info**\n\nIn the case of the ``on_exception`` decorator, all ``on_backoff`` and\n``on_giveup`` handlers are called from within the except block for the\nexception being handled. Therefore exception info is available to the\nhandler functions via the python standard library, specifically\n``sys.exc_info()`` or the ``traceback`` module.\n\nLogging configuration\n---------------------\n\nErrors and backoff and retry attempts are logged to the 'backoff'\nlogger. By default, this logger is configured with a NullHandler, so\nthere will be nothing output unless you configure a handler.\nProgrammatically, this might be accomplished with something as simple\nas::\n\n    logging.getLogger('backoff').addHandler(logging.StreamHandler())\n\nThe default logging level is ERROR, which corresponds to logging anytime\n``max_tries`` is exceeded as well as any time a retryable exception is\nraised. If you would instead like to log any type of retry, you can\nset the logger level to INFO::\n\n    logging.getLogger('backoff').setLevel(logging.INFO)", "description_content_type": null, "docs_url": null, "download_url": "", "downloads": {"last_day": -1, "last_month": -1, "last_week": -1}, "home_page": "https://github.com/a.sk/backoff-async", "keywords": "backoff function decorator async", "license": "MIT", "maintainer": "Alexandr Skurikhin", "maintainer_email": "a.skurihin@gmail.com", "name": "backoff-async", "package_url": "https://pypi.org/project/backoff-async/", "platform": "UNKNOWN", "project_url": "https://pypi.org/project/backoff-async/", "project_urls": {"Homepage": "https://github.com/a.sk/backoff-async"}, "release_url": "https://pypi.org/project/backoff-async/2.0.0/", "requires_dist": null, "requires_python": null, "summary": "Function decoration for backoff and retry async functions", "version": "2.0.0", "yanked": false, "html_description": "<div class=\"project-description\">\n            <div id=\"backoff\">\n<h2>backoff</h2>\n<a href=\"https://travis-ci.org/litl/backoff?branch=master\" rel=\"nofollow\"><img alt=\"https://travis-ci.org/litl/backoff.svg?branch=master\" src=\"https://warehouse-camo.ingress.cmh1.psfhosted.org/cb695a30cace15fdf6698c255c757ffbc593d324/68747470733a2f2f7472617669732d63692e6f72672f6c69746c2f6261636b6f66662e7376673f6272616e63683d6d6173746572\"></a>\n<a href=\"https://coveralls.io/r/litl/backoff?branch=master\" rel=\"nofollow\"><img alt=\"https://coveralls.io/repos/litl/backoff/badge.svg?branch=master\" src=\"https://warehouse-camo.ingress.cmh1.psfhosted.org/76492703c2e4a99e1826f462c716c8641e3bd9de/68747470733a2f2f636f766572616c6c732e696f2f7265706f732f6c69746c2f6261636b6f66662f62616467652e7376673f6272616e63683d6d6173746572\"></a>\n<p><strong>Function decoration for backoff and retry</strong></p>\n<p>This module provides function decorators which can be used to wrap a\nfunction such that it will be retried until some condition is met. It\nis meant to be of use when accessing unreliable resources with the\npotential for intermittent failures i.e. network resources and external\nAPIs. Somewhat more generally, it may also be of use for dynamically\npolling resources for externally generated content.</p>\n</div>\n<div id=\"examples\">\n<h2>Examples</h2>\n<p>Since Kenneth Reitz\u2019s <a href=\"http://python-requests.org\" rel=\"nofollow\">requests</a> module\nhas become a defacto standard for HTTP clients in python, networking\nexamples below are written using it, but it is in no way required by\nthe backoff module.</p>\n<div id=\"backoff-on-exception\">\n<h3>@backoff.on_exception</h3>\n<p>The <tt>on_exception</tt> decorator is used to retry when a specified exception\nis raised. Here\u2019s an example using exponential backoff when any\n<tt>requests</tt> exception is raised:</p>\n<pre>@backoff.on_exception(backoff.expo,\n                      requests.exceptions.RequestException,\n                      max_tries=8)\ndef get_url(url):\n    return requests.get(url)\n</pre>\n<p>The decorator will also accept a tuple of exceptions for cases where\nyou want the same backoff behavior for more than one exception type:</p>\n<pre>@backoff.on_exception(backoff.expo,\n                      (requests.exceptions.Timeout,\n                       requests.exceptions.ConnectionError),\n                      max_tries=8)\ndef get_url(url):\n    return requests.get(url)\n</pre>\n<p>In some cases the raised exception instance itself may need to be\ninspected in order to determine if it is a retryable condition. The\n<tt>giveup</tt> keyword arg can be used to specify a function which accepts\nthe exception and returns a truthy value if the exception should not\nbe retried:</p>\n<pre>def fatal_code(e):\n    return 400 &lt;= e.response.status_code &lt; 500\n\n@backoff.on_exception(backoff.expo,\n                      requests.exceptions.RequestException,\n                      max_tries=8,\n                      giveup=fatal_code)\ndef get_url(url):\n    return requests.get(url)\n</pre>\n</div>\n<div id=\"backoff-on-predicate\">\n<h3>@backoff.on_predicate</h3>\n<p>The <tt>on_predicate</tt> decorator is used to retry when a particular\ncondition is true of the return value of the target function.  This may\nbe useful when polling a resource for externally generated content.</p>\n<p>Here\u2019s an example which uses a fibonacci sequence backoff when the\nreturn value of the target function is the empty list:</p>\n<pre>@backoff.on_predicate(backoff.fibo, lambda x: x == [], max_value=13)\ndef poll_for_messages(queue):\n    return queue.get()\n</pre>\n<p>Extra keyword arguments are passed when initializing the\nwait generator, so the <tt>max_value</tt> param above is passed as a keyword\narg when initializing the fibo generator.</p>\n<p>When not specified, the predicate param defaults to the falsey test,\nso the above can more concisely be written:</p>\n<pre>@backoff.on_predicate(backoff.fibo, max_value=13)\ndef poll_for_message(queue)\n    return queue.get()\n</pre>\n<p>More simply, a function which continues polling every second until it\ngets a non-falsey result could be defined like like this:</p>\n<pre>@backoff.on_predicate(backoff.constant, interval=1)\ndef poll_for_message(queue)\n    return queue.get()\n</pre>\n</div>\n<div id=\"jitter\">\n<h3>Jitter</h3>\n<p>A jitter algorithm can be supplied with the <tt>jitter</tt> keyword arg to\neither of the backoff decorators. This argument should be a function\naccepting the original unadulterated backoff value and returning it\u2019s\njittered counterpart.</p>\n<p>As of version 1.2, the default jitter function <tt>backoff.full_jitter</tt>\nimplements the \u2018Full Jitter\u2019 algorithm as defined in the AWS\nArchitecture Blog\u2019s <a href=\"https://www.awsarchitectureblog.com/2015/03/backoff.html\" rel=\"nofollow\">Exponential Backoff And Jitter</a> post.</p>\n<p>Previous versions of backoff defaulted to adding some random number of\nmilliseconds (up to 1s) to the raw sleep value. If desired, this\nbehavior is now available as <tt>backoff.random_jitter</tt>.</p>\n</div>\n<div id=\"using-multiple-decorators\">\n<h3>Using multiple decorators</h3>\n<p>The backoff decorators may also be combined to specify different\nbackoff behavior for different cases:</p>\n<pre>@backoff.on_predicate(backoff.fibo, max_value=13)\n@backoff.on_exception(backoff.expo,\n                      requests.exceptions.HTTPError,\n                      max_tries=4)\n@backoff.on_exception(backoff.expo,\n                      requests.exceptions.TimeoutError,\n                      max_tries=8)\ndef poll_for_message(queue):\n    return queue.get()\n</pre>\n</div>\n<div id=\"runtime-configuration\">\n<h3>Runtime Configuration</h3>\n<p>The decorator functions <tt>on_exception</tt> and <tt>on_predicate</tt> are\ngenerally evaluated at import time. This is fine when the keyword args\nare passed as constant values, but suppose we want to consult a\ndictionary with configuration options that only become available at\nruntime. The relevant values are not available at import time. Instead,\ndecorator functions can be passed callables which are evaluated at\nruntime to obtain the value:</p>\n<pre>def lookup_max_tries():\n    # pretend we have a global reference to 'app' here\n    # and that it has a dictionary-like 'config' property\n    return app.config[\"BACKOFF_MAX_TRIES\"]\n\n@backoff.on_exception(backoff.expo,\n                      ValueError,\n                      max_tries=lookup_max_tries)\n</pre>\n<p>More cleverly, you might define a function which returns a lookup\nfunction for a specified variable:</p>\n<pre>def config(app, name):\n    return functools.partial(app.config.get, name)\n\n@backoff.on_exception(backoff.expo,\n                      ValueError,\n                      max_value=config(app, \"BACKOFF_MAX_VALUE\")\n                      max_tries=config(app, \"BACKOFF_MAX_TRIES\"))\n</pre>\n</div>\n<div id=\"event-handlers\">\n<h3>Event handlers</h3>\n<p>Both backoff decorators optionally accept event handler functions\nusing the keyword arguments <tt>on_success</tt>, <tt>on_backoff</tt>, and <tt>on_giveup</tt>.\nThis may be useful in reporting statistics or performing other custom\nlogging.</p>\n<p>Handlers must be callables with a unary signature accepting a dict\nargument. This dict contains the details of the invocation. Valid keys\ninclude:</p>\n<ul>\n<li><em>target</em>: reference to the function or method being invoked</li>\n<li><em>args</em>: positional arguments to func</li>\n<li><em>kwargs</em>: keyword arguments to func</li>\n<li><em>tries</em>: number of invocation tries so far</li>\n<li><em>wait</em>: seconds to wait (<tt>on_backoff</tt> handler only)</li>\n<li><em>value</em>: value triggering backoff (<tt>on_predicate</tt> decorator only)</li>\n</ul>\n<p>A handler which prints the details of the backoff event could be\nimplemented like so:</p>\n<pre>def backoff_hdlr(details):\n    print (\"Backing off {wait:0.1f} seconds afters {tries} tries \"\n           \"calling function {func} with args {args} and kwargs \"\n           \"{kwargs}\".format(**details))\n\n@backoff.on_exception(backoff.expo,\n                      requests.exceptions.RequestException,\n                      on_backoff=backoff_hdlr)\ndef get_url(url):\n    return requests.get(url)\n</pre>\n<p><strong>Multiple handlers per event type</strong></p>\n<p>In all cases, iterables of handler functions are also accepted, which\nare called in turn. For example, you might provide a simple list of\nhandler functions as the value of the <tt>on_backoff</tt> keyword arg:</p>\n<pre>@backoff.on_exception(backoff.expo,\n                      requests.exceptions.RequestException,\n                      on_backoff=[backoff_hdlr1, backoff_hdlr2])\ndef get_url(url):\n    return requests.get(url)\n</pre>\n<p><strong>Getting exception info</strong></p>\n<p>In the case of the <tt>on_exception</tt> decorator, all <tt>on_backoff</tt> and\n<tt>on_giveup</tt> handlers are called from within the except block for the\nexception being handled. Therefore exception info is available to the\nhandler functions via the python standard library, specifically\n<tt>sys.exc_info()</tt> or the <tt>traceback</tt> module.</p>\n</div>\n<div id=\"logging-configuration\">\n<h3>Logging configuration</h3>\n<p>Errors and backoff and retry attempts are logged to the \u2018backoff\u2019\nlogger. By default, this logger is configured with a NullHandler, so\nthere will be nothing output unless you configure a handler.\nProgrammatically, this might be accomplished with something as simple\nas:</p>\n<pre>logging.getLogger('backoff').addHandler(logging.StreamHandler())\n</pre>\n<p>The default logging level is ERROR, which corresponds to logging anytime\n<tt>max_tries</tt> is exceeded as well as any time a retryable exception is\nraised. If you would instead like to log any type of retry, you can\nset the logger level to INFO:</p>\n<pre>logging.getLogger('backoff').setLevel(logging.INFO)\n</pre>\n</div>\n</div>\n\n          </div>"}, "last_serial": 2363731, "releases": {"2.0.0": [{"comment_text": "", "digests": {"md5": "b185637881ab05e479b48c3b7c10d3c8", "sha256": "953254ed20c8743b4358df9c0b4fd09d59ff1882bb4725848b9e80ff5820c4fe"}, "downloads": -1, "filename": "backoff-async-2.0.0.tar.gz", "has_sig": false, "md5_digest": "b185637881ab05e479b48c3b7c10d3c8", "packagetype": "sdist", "python_version": "source", "requires_python": null, "size": 7255, "upload_time": "2016-09-26T10:04:01", "upload_time_iso_8601": "2016-09-26T10:04:01.549230Z", "url": "https://files.pythonhosted.org/packages/e4/36/a3c9dc5a3ba3c6c40c6b199957a60933fa2b82aebbaed49443ca23657a50/backoff-async-2.0.0.tar.gz", "yanked": false}]}, "urls": [{"comment_text": "", "digests": {"md5": "b185637881ab05e479b48c3b7c10d3c8", "sha256": "953254ed20c8743b4358df9c0b4fd09d59ff1882bb4725848b9e80ff5820c4fe"}, "downloads": -1, "filename": "backoff-async-2.0.0.tar.gz", "has_sig": false, "md5_digest": "b185637881ab05e479b48c3b7c10d3c8", "packagetype": "sdist", "python_version": "source", "requires_python": null, "size": 7255, "upload_time": "2016-09-26T10:04:01", "upload_time_iso_8601": "2016-09-26T10:04:01.549230Z", "url": "https://files.pythonhosted.org/packages/e4/36/a3c9dc5a3ba3c6c40c6b199957a60933fa2b82aebbaed49443ca23657a50/backoff-async-2.0.0.tar.gz", "yanked": false}], "timestamp": "Thu May  7 18:15:08 2020"}